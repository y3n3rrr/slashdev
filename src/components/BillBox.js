import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  ScrollView,
  Dimensions,
  TouchableOpacity,
} from 'react-native';
import {appColors} from '../utils';
import {Box, Text as NText, HStack, Button} from 'native-base';
import Icon from 'react-native-vector-icons/FontAwesome';

export const BillBox = props => {
  const {subTotal, deliveryFee, discount, total} = props;
  return (
    <Box
      safeArea
      flex={1}
      bg="white"
      borderRadius={20}
      _text={{
        fontSize: 'md',
        fontWeight: 'bold',
        color: 'black',
      }}
    >
      <View style={{paddingLeft: 30, paddingRight: 30}}>
        <Text style={styles.HEADER_TEXT}>Have a coupon code? Enter here</Text>
        <Box
          justifyContent="center"
          borderWidth={0.2}
          borderColor="gray.400"
          bg="white"
          borderRadius={8}
          flexDirection="column"
          p={4}
          top={5}
          mb={10}
          _text={{
            fontSize: 'md',
            fontWeight: 'bold',
            color: 'black',
          }}
        >
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              alignItems: 'center',
            }}
          >
            <Text style={styles.TEXT_COUPON}>XXZXEWT</Text>
            <View
              style={{
                flexDirection: 'row',
                justifyContent: 'space-between',
                alignItems: 'center',
              }}
            >
              <Text style={styles.TEXT_HEADER}>Available </Text>
              <Icon
                name="check-circle"
                size={20}
                color={appColors.plantGreen}
              />
            </View>
          </View>
        </Box>
        <View>
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              alignItems: 'center',
            }}
          >
            <Text style={styles.PRICE_LABEL}>Sub Total</Text>
            <Text style={styles.PRICE_LABEL}>${subTotal}</Text>
          </View>
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              alignItems: 'center',
              top: 15,
            }}
          >
            <Text style={styles.PRICE_LABEL}>Delivery Fee:</Text>
            <Text style={styles.PRICE_LABEL}>${deliveryFee}</Text>
          </View>
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              alignItems: 'center',
              top: 30,
            }}
          >
            <Text style={styles.PRICE_LABEL}>Discount:</Text>
            <Text style={styles.PRICE_LABEL}>${discount?.toFixed(2)}</Text>
          </View>
        </View>
      </View>

      <View
        style={{
          flex: 1,
          justifyContent: 'flex-end',
          alignItems: 'center',
          bottom: 10,
          paddingLeft: 30,
          paddingRight: 30,
        }}
      >
        <Text
          ellipsizeMode="clip"
          numberOfLines={1}
          style={{bottom: 30, color: 'gray'}}
        >
          - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
          - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
          - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
          - - - - - - - - - - - - - - - - - - - - - - -
        </Text>
        <View
          style={{
            width: '100%',
            flexDirection: 'row',
            justifyContent: 'space-between',
            bottom: 15,
          }}
        >
          <Text style={styles.PRICE_LABEL}>Total:</Text>
          <Text style={styles.PRICE_LABEL_TOTAL}>${total?.toFixed(2)}</Text>
        </View>
        <TouchableOpacity
          style={{
            borderRadius: 30,
            backgroundColor: '#4DD89F',
            width: '100%',
            paddingLeft: 3,
            paddingRight: 5,
          }}
          onPress={() => {}}
        >
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              alignItems: 'center',
            }}
          >
            <Icon name="check-circle" size={36} color="white" />
            <Text style={{color: 'white', fontWeight: '600'}}>Checkout</Text>
            <Icon
              name="angle-double-right"
              size={16}
              style={{paddingRight: 15, height: 0}}
              color="white"
            />
          </View>
        </TouchableOpacity>
      </View>
    </Box>
  );
};
const styles = StyleSheet.create({
  HEADER_TEXT: {
    fontSize: 14,
    fontWeight: '600',
  },
  TEXT_COUPON: {
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
    paddingRight: 15,
  },
  TEXT_HEADER: {
    fontSize: 16,
    fontWeight: '600',
    textAlign: 'center',
    paddingRight: 15,
    color: appColors.plantGreen,
  },
  TEXT_SUBTEXT: {
    paddingTop: 25,
    paddingBottom: 25,
    textAlign: 'center',
    color: 'darkgray',
    opacity: 0.8,
  },
  PRICE_LABEL: {
    fontSize: 15,
    color: 'gray',
  },
  PRICE_LABEL_TOTAL: {
    fontSize: 16,
    fontWeight: '800',
    color: appColors.plantGreen,
  },
});
