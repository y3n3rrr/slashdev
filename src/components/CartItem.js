import React, {useContext, useEffect} from 'react';

import {
  ImageBackground,
  StyleSheet,
  Text,
  useColorScheme,
  View,
  TouchableOpacity,
} from 'react-native';

import Icon from 'react-native-vector-icons/FontAwesome';
import {Box, Badge, Image} from 'native-base';
import {REMOVE_PRODUCT_FROM_CART, ProductContext} from '../context';

export const CartItem = props => {
  const {id, title, price, imageUrl, count, onPressIncrement} = props;
  const {state, dispatch} = useContext(ProductContext);
  useEffect(() => {
    if (count <= 0) {
      dispatch({type: REMOVE_PRODUCT_FROM_CART, payload: {id}});
    }
  }, [count]);
  return (
    <Box
      bg="white"
      borderRadius={20}
      flexDirection="column"
      p={5}
      m={5}
      _text={{
        fontSize: 'md',
        fontWeight: 'bold',
        color: 'black',
      }}
    >
      <View style={{flexDirection: 'row'}}>
        <Image size="xs" source={{uri: imageUrl}} alt="No Preview Image" />
        <View
          style={{
            flex: 1,
            flexDirection: 'row',
            justifyContent: 'space-between',
            alignItems: 'center',
          }}
        >
          <View>
            <Text>{title}</Text>
            <Text>${price.toFixed(2)}</Text>
          </View>
          <View style={{flexDirection: 'row'}}>
            <TouchableOpacity
              onPress={() => onPressIncrement(id, 1)}
              style={styles.roundButton}
            >
              <Icon name="plus" size={8} color="white" />
            </TouchableOpacity>
            <Text style={styles.COUNT_TEXT}>{count}</Text>
            <TouchableOpacity
              onPress={() => onPressIncrement(id, -1)}
              style={styles.roundButton}
            >
              <Icon name="minus" size={6} color="white" />
            </TouchableOpacity>
          </View>
        </View>
      </View>
    </Box>
  );
};
const styles = StyleSheet.create({
  COUNT_TEXT: {
    fontWeight: '700',
    marginLeft: 10,
    marginRight: 10,
  },
  TEXT_HEADER: {
    fontSize: 30,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  TEXT_SUBTEXT: {
    paddingTop: 25,
    paddingBottom: 25,
    textAlign: 'center',
    color: 'darkgray',
    opacity: 0.8,
  },
  roundButton: {
    width: 16,
    height: 16,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 100,
    backgroundColor: '#4DD89F',
  },
});
