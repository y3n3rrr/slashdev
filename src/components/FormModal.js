// original source: https://medium.com/@ndyhrdy/making-the-bottom-sheet-modal-using-react-native-e226a30bed13 ðŸ™‡
import React, {useEffect} from 'react';
import {ImageBackground, StyleSheet, Text} from 'react-native';
import {
  Image,
  Button,
  Modal,
  Stack,
  FormControl,
  Input,
  Center,
} from 'native-base';
import {useForm, Controller} from 'react-hook-form';
import {SSRProvider} from '@react-aria/ssr';
import {appColors} from '../utils';
export const FormModal = props => {
  const {visible, onDismiss, onSubmit} = props;
  const {
    control,
    handleSubmit,
    formState: {errors},
    reset,
  } = useForm();

  return (
    <Modal
      isOpen={visible}
      onClose={onDismiss}
      safeAreaTop={true}
      avoidKeyboard
    >
      <Modal.Content maxWidth="350">
        <Modal.CloseButton onPress={onDismiss} />
        <Modal.Header>Add New Product</Modal.Header>
        <Modal.Body>
          <SSRProvider>
            <FormControl isRequired>
              <FormControl.Label>Title</FormControl.Label>
              <Controller
                control={control}
                render={({field: {onChange, onBlur, value}}) => (
                  <Input
                    onBlur={onBlur}
                    placeholder="Enter Title"
                    onChangeText={val => onChange(val)}
                    value={value}
                  />
                )}
                name="title"
                rules={{
                  required: 'Title is required',
                  minLength: {
                    value: 3,
                    message: 'Title should be at least 3 character',
                  },
                }}
                defaultValue=""
              />
              {errors?.title?.message && (
                <Text style={styles.ERROR_TEXT}>{errors?.title?.message}</Text>
              )}
            </FormControl>
            <FormControl mt="3" isRequired>
              <FormControl.Label>Price</FormControl.Label>
              <Controller
                control={control}
                render={({field: {onChange, onBlur, value}}) => (
                  <Input
                    value={value}
                    onBlur={onBlur}
                    placeholder="Enter Price"
                    keyboardType="numeric"
                    onChangeText={val => {
                      if (!isNaN(val)) {
                        onChange(val);
                      }
                    }}
                  />
                )}
                name="price"
                rules={{
                  required: 'Price is required',
                  min: {
                    value: 0.000001,
                    message: 'Enter valid price value',
                  },
                }}
                defaultValue=""
              />
              {errors?.price?.message && (
                <Text style={styles.ERROR_TEXT}>{errors?.price?.message}</Text>
              )}
            </FormControl>
            <FormControl mt="3">
              <FormControl.Label>Description</FormControl.Label>
              <Controller
                control={control}
                render={({field: {onChange, onBlur, value}}) => (
                  <Input
                    onBlur={onBlur}
                    onChangeText={val => onChange(val)}
                    value={value}
                  />
                )}
                name="description"
                defaultValue=""
              />
            </FormControl>
            <FormControl mt="3">
              <FormControl.Label>Image Url</FormControl.Label>
              <Controller
                control={control}
                render={({field: {onChange, onBlur, value}}) => (
                  <Input
                    onBlur={onBlur}
                    onChangeText={val => onChange(val)}
                    value={value}
                  />
                )}
                name="imageUrl"
                defaultValue=""
              />
            </FormControl>
          </SSRProvider>
        </Modal.Body>
        <Modal.Footer>
          <Button.Group space={2}>
            <Button variant="ghost" colorScheme="blueGray" onPress={onDismiss}>
              Cancel
            </Button>
            <Button
              onPress={handleSubmit(formValues => {
                onSubmit(formValues);
                onDismiss();
                reset();
              })}
            >
              Save
            </Button>
          </Button.Group>
        </Modal.Footer>
      </Modal.Content>
    </Modal>
  );
};

const styles = StyleSheet.create({
  ERROR_TEXT: {
    top: 5,
    fontSize: 12,
    color: appColors.errorTextColor,
  },
});
