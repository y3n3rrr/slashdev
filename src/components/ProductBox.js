import React from 'react';
import {Text, View, StyleSheet} from 'react-native';
import {Badge, Image} from 'native-base';
import {appColors} from '../utils';

export const ProductBox = props => {
  const {id, title, description, price, imageUrl} = props;
  return (
    <View style={{flexDirection: 'row', padding: 15}} key={`Product_${id}`}>
      <Image size="sm" source={{uri: imageUrl}} alt="No Preview Image" />
      <View style={styles.CONTAINER}>
        <View>
          <Text style={styles.TITLE}>{title}</Text>
          <Text style={styles.SUBTEXT}>{description}</Text>
        </View>
        <Badge
          style={{borderRadius: 24, backgroundColor: appColors.plantGreen}}
          _text={{
            color: 'white',
            fontWeight: '700',
          }}
        >
          {`$${price.toFixed(2)}`}
        </Badge>
      </View>
    </View>
  );
};
const styles = StyleSheet.create({
  CONTAINER: {
    flex: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'row',
    paddingLeft: 10,
  },
  IMAGE: {
    flex: 1,
    justifyContent: 'center',
  },
  TITLE: {
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  SUBTEXT: {
    color: 'darkgray',
    opacity: 0.9,
    fontSize: 12,
  },
});
