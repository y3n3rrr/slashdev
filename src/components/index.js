export * from './BottomModal';
export * from './FormModal';
export * from './CartItem';
export * from './ProductBox';
export * from './BillBox';
