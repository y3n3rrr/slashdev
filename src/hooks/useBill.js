import React, { useContext, useEffect, useState } from "react";

import { ProductContext } from "../context/ProductContext";

export default function useBillCalculator() {
  const { state } = useContext(ProductContext);
  const { Products, ShoppingCart } = state;
  const [bill, setBill] = React.useState(false);

  const getBill = () => {
    const result = calculateBill(ShoppingCart, Products);
    setBill(result);
  };

  return [bill, getBill];
}

export const calculateBill = (cartItems, productItems) => {
  let subTotal = 0;
  let discount = 0;
  const deliveryFee = 5;

  cartItems.map((item) => {
    let dicountRate = item.count >= 3 ? 0.5 : 1;
    const product = productItems.filter((f) => f.id === item.id)[0];
    subTotal += product.price * item.count;
    if (dicountRate < 1) discount += product.price * item.count * dicountRate;
  });
  let total = subTotal + deliveryFee - discount;

  return { subTotal, deliveryFee, discount, total };
};
