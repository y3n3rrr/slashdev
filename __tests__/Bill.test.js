import { renderHook } from "@testing-library/react-hooks";
import { calculateBill } from "../src/hooks/useBill";
import { ProductBox } from "../src/components/ProductBox";
import renderer from "react-test-renderer";
import { NativeBaseProvider } from "native-base";

it("Bill calculator", () => {
  const productItems = [
    { id: 1, price: 10 },
    { id: 2, price: 20 },
    { id: 3, price: 50 },
  ];
  const cartItems = [
    { id: 1, count: 1 },
    { id: 2, count: 5 },
  ];
  const result = calculateBill(cartItems, productItems);
  expect(result).toStrictEqual({
    deliveryFee: 5,
    discount: 50,
    subTotal: 110,
    total: 65,
  });
});

const data = {
  id: 1,
  title: "title",
  description: "description",
  price: 12,
  imageUrl: "",
};
it("ProductBox", () => {
  const tree = renderer
    .create(
      <NativeBaseProvider>
        <ProductBox {...data} />
      </NativeBaseProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
