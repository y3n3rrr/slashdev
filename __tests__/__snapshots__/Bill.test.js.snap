// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductBox 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;

exports[`ProductBox test 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;

exports[`check state after process 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "padding": 15,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingLeft": 10,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        title
      </Text>
      <Text
        style={
          Object {
            "color": "darkgray",
            "fontSize": 12,
            "opacity": 0.9,
          }
        }
      >
        description
      </Text>
    </View>
    &gt;
    $12.00
  </View>
</View>
`;
